packages/runtime-common/realm-auth-client.ts

```diff
@@ -0,0 +1,111 @@
+import { MatrixClient } from './matrix-client';
+
+export class RealmAuthClient {
+  private realmURL: URL;
+  private jwt: string | undefined;
+  private matrixClient: MatrixClient;
+
+  constructor(
+    matrixUsername: string,
+    matrixPassword: string,
+    matrixURL: URL,
+    realmURL: URL,
+  ) {
+    this.realmURL = realmURL;
+    this.matrixClient = new MatrixClient(
+      matrixURL,
+      matrixUsername,
+      matrixPassword,
+    );
+  }
+
+  async getJWT() {
+    let extraPeriodBeforeExpirySeconds = 60;
+    let jwt: string;
+
+    if (!this.jwt) {
+      jwt = await this.createRealmSession();
+      this.jwt = jwt;
+      return jwt;
+    } else {
+      let jwtData = JSON.parse(atob(this.jwt.split('.')[1]));
```

Comment:
Not so much a performance thing because I expect this is relatively fast, but for clarity it might make sense to have JWT be structured data that we parse when it's created (and check the format is correct) rather than split/unbase64/json each time we want to check the expiry. Partly also to have a place to reference a couple of sources, I'm not used to the structure so went to check that exp was seconds not milliseconds, that kind of thing.

We may want to allow for exp to be missing as it's not a required part of the JWT spec.